codec-http/src/main/java/io/netty/handler/codec/http/HttpKeepAliveHandler.java

```diff
@@ -0,0 +1,132 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http;
+
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.ChannelPromise;
+import io.netty.util.Attribute;
+import io.netty.util.AttributeKey;
+
+import static io.netty.handler.codec.http.HttpUtil.*;
+import static java.lang.Boolean.*;
+
+/**
+ * HttpKeepAliveHandler helps maintain persistent connections when requested.
+ * <p>
+ * When a persistent connection request arrives then HttpKeepAliveHandler will make sure that the response also contains
+ * the proper 'Connection' header.  The channel will automatically be closed for any LastHttpContent being written
+ * through this handler, iff the connection cannot be persisted.
+ * <p>
+ * Since {@link HttpKeepAliveHandler} expects {@link HttpObject}s it should be added after {@link HttpServerCodec} but
+ * before any other handlers that might send a {@link HttpResponse}.
+ * <blockquote>
+ * <pre>
+ *  {@link ChannelPipeline} p = ...;
+ *  ...
+ *  p.addLast("serverCodec", new {@link HttpServerCodec}());
+ *  p.addLast("httpKeepAlive", <b>new {@link HttpKeepAliveHandler}()</b>);
+ *  p.addLast("aggregator", new {@link HttpObjectAggregator}(1048576));
+ *  ...
+ *  p.addLast("handler", new HttpRequestHandler());
+ *  </pre>
+ * </blockquote>
+ */
+@ChannelHandler.Sharable
+public class HttpKeepAliveHandler extends ChannelDuplexHandler {
+
+    private static final String MULTIPART_PREFIX = "multipart";
+
+    /**
+     * Will close the channel if the connection couldn't be persisted.
+     *
+     * Only works if HttpKeepAliveHandler is part of the pipeline.
+     */
+    public static final ChannelFutureListener CLOSE_UNLESS_KEEP_ALIVE = new ChannelFutureListener() {
+        @Override
+        public void operationComplete(ChannelFuture channelFuture) throws Exception {
+            Boolean keepAlive = channelFuture.channel().attr(KEEP_ALIVE_ATTR).get();
+            if (!TRUE.equals(keepAlive) && channelFuture.channel().isActive()) {
+                channelFuture.channel().close();
+            }
+        }
+    };
+
+    private static final AttributeKey<Boolean> KEEP_ALIVE_ATTR =
+            AttributeKey.valueOf(HttpKeepAliveHandler.class, "KEEP_ALIVE_ATTR");
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        // read message and track if it was keepAlive
+        if (msg instanceof HttpRequest) {
+            HttpRequest request = (HttpRequest) msg;
+            ctx.channel().attr(KEEP_ALIVE_ATTR).set(HttpUtil.isKeepAlive(request));
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Unnecessary use of fully qualified name 'HttpUtil.isKeepAlive' due to existing static import 'io.netty.handler.codec.http.HttpUtil.*' [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUnnecessaryFullyQualifiedName)